Add https support for git operations

juliogonzalez opened the pull request:
HTTPS can be enabled by using the switch --https (otherwise the gitbot will use SSH as it does currently).

This is useful to avoid any SSH configuration when using GitBot with public projects.

juliogonzalez commented on lib/git_op.rb:
```diff
@@ -17,19 +17,24 @@ def initialize(git_dir, pr)
   end
 
-  def ck_or_clone_git(repo)
+  def ck_or_clone_git(repo, https = False)
```
Default value is added here, so the current test at tests/t_git_operation.rb can work as it is.

Not sure if we should test there by calling ck_or_clone with https = False and https = True, instead of using the manual test at tests/manual.sh

juliogonzalez commented:
Ok, fixing style problems.

juliogonzalez commented on lib/gitbot_backend.rb:
```diff
@@ -35,17 +35,18 @@ def run_script
   end # main function for doing the test
 
-  def pr_test(upstream, pr_sha_com, repo, pr_branch, pr)
+  def pr_test(repo, pr, https)
```
Refactored to avoid problems at rubocop: https://travis-ci.org/openSUSE/gitbot/builds/278835768?utm_source=github_status&utm_medium=notification

MalloZup replied:
+1

juliogonzalez commented on lib/gitbot_backend.rb:
```diff
@@ -35,24 +35,20 @@ def run_script
     git = GitOp.new(@git_dir, pr)
-    # get author:
-    pr_com = @client.commit(repo, pr_sha_com)
-    _author_pr = pr_com.author.login
```
Removed, since it was not used.

MalloZup replied:
good

juliogonzalez commented on lib/gitbot_backend.rb:
```diff
@@ -35,24 +35,20 @@ def run_script
   def magic_comment(repo, pr_num)
-    comments = @client.issue_comments(repo, pr_num)
-    comments.each do |com|
+    @client.issue_comments(repo, pr_num).each do |com|
```
Helps preventing: ```C: Assignment Branch Condition size for pr_test is too high.```

https://travis-ci.org/openSUSE/gitbot/builds/278839053

juliogonzalez commented:
Do not merge yet, checkout_into() is still using SSH

juliogonzalez commented:
Verified and ready for review.

With --https:
```
$ git remote -v
origin  https://github.com/cd-training/ebs-tools.git (fetch)
origin  https://github.com/cd-training/ebs-tools.git (push)
remtest-py      https://github.com/juliogonzalez/ebs-tools (fetch)
remtest-py      https://github.com/juliogonzalez/ebs-tools (push)
```

Without it:
```
$ git remote -v
origin  git@github.com:cd-training/ebs-tools.git (fetch)
origin  git@github.com:cd-training/ebs-tools.git (push)
remtest-py      git@github.com:juliogonzalez/ebs-tools.git (fetch)
remtest-py      git@github.com:juliogonzalez/ebs-tools.git (push)
```

In any case, a refactor for the git operations will be needed since it is possible to fetch commits from PRs without adding extra remotes.

juliogonzalez commented on lib/git_op.rb:
```diff
@@ -17,19 +17,24 @@ def initialize(git_dir, pr)
     return if File.directory?(git_dir)
     FileUtils.mkdir_p(git_dir)
     Dir.chdir git_dir
-    puts `git clone git@github.com:#{repo}.git`
```
Why? Lines 24-28 already define the repository URL correctly depending on the presence (or not) of --https parameter.

juliogonzalez commented:
@MalloZup why do we add an extra dependency? We do not need any new library, since we are using git commands.

For sure we are adding new parameters, but I think this is something that git bot is going to need to improve it (I already thinking about two more, so we can correctly support webhooks)

Adding new parameters is not a problem as soon as we can test them (so far the test is already present) and are documented. A lot of CLI programs have a lot of different options (have a look at tar, for example) and it is not a problem. And in this case is not even a mandatory parameter which requires existing users to read the docs (it's fully retrocompatible)

In the end the user should be able to choose the protocol he wants to use to avoid wasting time. As an example there could be a firewall blocking SSH traffic having the connection hang for a couple of minutes. Or he could have 5 ssh keys at the ssh-agent, not valid for GitHub, retrying five times before using https and therefore wasting time.

Of course, to make the configuration easier we should consider enabling configuration files in the future so the user can just run ./gitbot.rb --config <myconfigfile> and avoid using parameters (or at least most of them).

MalloZup commented:
Ok, this is a good point, then . +1 I will review deeply it then on monday, but so far its good

juliogonzalez commented:
Good, ping if you find anything strange :-)

By the way, I am not 100% if we should call the parameter `--https` or `--use-https` so if you prefer to change it just let me know and I will do it quickly with sed.

MalloZup commented on lib/git_op.rb:
```diff
@@ -17,19 +17,24 @@ def initialize(git_dir, pr)
     FileUtils.mkdir_p(git_dir)
     Dir.chdir git_dir
-    puts `git clone git@github.com:#{repo}.git`
+    repo_url = if https
```
Here is better to have
```if_this_is_a_true_value ? then_the_result_is_nthis : else_it_is_this```
https://stackoverflow.com/questions/4252936/how-do-i-use-the-conditional-operator-in-ruby

juliogonzalez replied:
Ok, since the ternary operator won't fit in to a line with less than 80 characters, and I can't split the line since ternary operators should not use more than one line, that forces me to setup the variable twice:

* One time to store "git@github.com:" or "https://github.com/"
* Another time to add the project name, repo name and '.git'

Let me give it a try and will amend the commit.

juliogonzalez replied:
Fixed

MalloZup commented on lib/git_op.rb:
```diff
@@ -103,8 +108,13 @@ def branch_rem
 
-  def add_remote(rem_repo)
-    puts `git remote add #{rem_repo} #{pr.head.repo.ssh_url}`
+  def add_remote(rem_repo, https)
```
we can refactor this with the trynary op

MalloZup commented:
i think additionally that instead of having the `https` var as argument for method,

we should inject the @options to the constructor of git_op, in this way can control more and this is much more maintenable, and allow flexibility for future.

https://github.com/openSUSE/gitbot/blob/master/lib/gitbot_backend.rb#L48
````(  git = GitOp.new(@git_dir, pr, @options)```

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun @changelog_shouldpass

MalloZup commented:
@gitbot changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun check-option-test !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun check-option-test !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun check-option-test !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun check-option-test !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun check-option-test !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun check-option-test !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun check-option-test !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

MalloZup commented:
@gitbot rerun check-option-test !!!

MalloZup commented:
@gitbot rerun check-option-test !!!

MalloZup commented:
no changelog needed!

MalloZup commented:
@gitbot rerun changelog_shouldpass !!!

juliogonzalez commented:
@MalloZup, I fully agree with your suggestion, since most probably we we'll need to add even more options.

Let's address this on a separate PR to refactor git_opt.rb, since we should also avoid having put()'s all over the code and instead create functions to have 'git clone' or 'or git pull' only once and call them when needed.

juliogonzalez commented:
Ok, it seems that:

https://travis-ci.org/openSUSE/gitbot/builds/279237788
```
1.59s$ rake
rubocop 
Inspecting 11 files
....C......
Offenses:
lib/gitbot_backend.rb:10:1: C: Class has too many lines. [158/152]
class GitbotBackend ...
```
Is triggered after b9950ccc4371eabe7512a5e86f6e13b95813e571, since at 'lib/gitbot_backend.rb' I did not added any new lines (in fact I removed one).

In fact master is already failing:
```
$ rubocop.ruby2.1 --auto-gen-config lib/gitbot_backend.rb 
Inspecting 1 file
C

Offenses:

lib/gitbot_backend.rb:10:1: C: Class has too many lines. [161/100]
class GitbotBackend ...
^^^^^^^^^^^^^^^^^^^

1 file inspected, 1 offense detected
```

@MalloZup could we ignore this failure for this PR and then refactor lib/gitbot_backend.rb?

MalloZup commented:
@juliogonzalez yop ignore rubocop, but i would inject the options to gitop class, so we make https less invasive to othrers methods.

juliogonzalez commented:
@MalloZup: refactored.

I also removed all references to ```repo``` at ```lib/gitbot_backend.rb``` and ```lib/git_op``` and replaced with ```@options[:repo]``` where needed.

juliogonzalez commented:
Please do not merge yet, I see that the acceptance tests were changed and we lost the test for https.

juliogonzalez commented:
Ready (I hope). But most probably we should refactor the acceptance tests.

MalloZup commented:
yop for the acceptance test, we will use rspec, this is more professional, and a refactor is always good of course. thx for the great pr
